File: code-smells-project/internal-tests.py
#!/usr/bin/env python3
"""
internal-tests.py -- checagens de caixa-branca, em processo, sem servidor.

Completa os outros dois validadores do projeto:
    arch-check.sh      -> estrutura (AP-16: rota nao toca persistencia)
    manual-tests.sh    -> comportamento via HTTP (precisa da app de pe)
    internal-tests.py  -> propriedades que a resposta HTTP nao revela

O que so da para provar aqui:
    1. Que get_todos (AP-08) nao volta a ser N+1. O numero de queries emitidas
       nao pode crescer com a quantidade de pedidos -- _hydrate_pedidos busca
       todos os itens/produtos de uma vez com WHERE ... IN (...).
    2. Que a checagem acima sabe reprovar (controle negativo). Um teste que so
       sabe passar nao mede nada: uma query por item (jeito antigo) TEM que
       crescer com N.

Uso:  python3 internal-tests.py
Saida: 0 = todas passaram, 1 = alguma falhou.
"""
import os
import sys
from contextlib import contextmanager

os.environ.setdefault("DB_PATH", ":memory:")

from app import create_app  # noqa: E402
from models import order_model  # noqa: E402
from models.db import get_db, _create_schema, _seed_if_empty  # noqa: E402

results = []


def check(name, passed, detail=""):
    results.append((name, passed))
    print(f"  {'✓' if passed else '✗'} {name}")
    if detail:
        print(f"      {detail}")


@contextmanager
def count_queries(db):
    """Conta quantos statements SQL rodam na conexao `db` durante o bloco `with`,
    via sqlite3.Connection.set_trace_callback (Cursor.execute e imutavel em CPython,
    nao da para monkeypatchar)."""
    count = 0

    def trace(_statement):
        nonlocal count
        count += 1

    db.set_trace_callback(trace)
    try:
        yield lambda: count
    finally:
        db.set_trace_callback(None)


def seed_pedido(db, usuario_id, itens):
    """Insere um pedido com N itens direto via SQL, sem passar por order_model.criar
    (que teria sua propria contagem de queries)."""
    cursor = db.cursor()
    cursor.execute(
        "INSERT INTO pedidos (usuario_id, status, total) VALUES (?, 'pendente', 0)",
        (usuario_id,),
    )
    pedido_id = cursor.lastrowid
    for produto_id, quantidade in itens:
        cursor.execute(
            """INSERT INTO itens_pedido (pedido_id, produto_id, quantidade, preco_unitario)
               VALUES (?, ?, ?, 1.0)""",
            (pedido_id, produto_id, quantidade),
        )
    db.commit()
    return pedido_id


def check_get_todos_query_count_constant(db):
    """1. get_todos: numero de queries constante, nao cresce com N pedidos."""
    produto_ids = [row["id"] for row in db.execute("SELECT id FROM produtos").fetchall()]

    for _ in range(3):
        seed_pedido(db, 1, [(produto_ids[0], 1), (produto_ids[1], 2)])
    with count_queries(db) as small_counter:
        small_result = order_model.get_todos(page=1, per_page=50)
    small_count = small_counter()

    for _ in range(12):
        seed_pedido(db, 1, [(produto_ids[0], 1), (produto_ids[1], 2)])
    with count_queries(db) as large_counter:
        large_result = order_model.get_todos(page=1, per_page=50)
    large_count = large_counter()

    check(
        "get_todos: numero de queries nao cresce com a quantidade de pedidos",
        small_count == large_count and large_count <= 2,
        f"{len(small_result)} pedido(s) -> {small_count} querie(s) | "
        f"{len(large_result)} pedido(s) -> {large_count} querie(s)",
    )


def check_naive_per_item_loop_would_grow(db):
    """2. Controle negativo: uma query por item (jeito antigo) TEM que crescer com N."""
    cursor = db.cursor()
    cursor.execute("SELECT id FROM itens_pedido")
    item_ids = cursor.fetchall()

    with count_queries(db) as counter:
        for _ in item_ids:
            cursor.execute("SELECT nome FROM produtos WHERE id = 1")
    naive_count = counter()

    check(
        "controle negativo: uma query por item cresce com N "
        "(prova que a checagem acima mede algo)",
        naive_count == len(item_ids),
        f"{len(item_ids)} item(ns) -> {naive_count} querie(s) no jeito ingenuo",
    )


def main():
    print("=" * 58)
    print("internal-tests -- code-smells-project")
    print("Checagens em processo (sem servidor)")
    print("=" * 58)
    print()

    app = create_app()

    # Um unico app context para toda a rodada: DB_PATH=:memory: vive e morre com
    # a conexao guardada em flask.g, entao contextos separados veriam bancos
    # distintos (e vazios) um do outro.
    with app.app_context():
        db = get_db()
        _create_schema()
        _seed_if_empty()
        check_get_todos_query_count_constant(db)
        check_naive_per_item_loop_would_grow(db)

    failed = [name for name, passed in results if not passed]
    print()
    print("=" * 58)
    if not failed:
        print(f"PASS: {len(results)} checagem(ns) internas OK.")
        print("=" * 58)
        sys.exit(0)
    print(f"FAIL: {len(failed)} de {len(results)} checagem(ns) falharam.")
    for name in failed:
        print(f"  - {name}")
    print("=" * 58)
    sys.exit(1)


if __name__ == "__main__":
    main()

File: ecommerce-api-legacy/internal-tests.js
#!/usr/bin/env node
//
// internal-tests.js — checagens de caixa-branca, em processo, sem servidor.
//
// Completa os outros dois validadores do projeto:
//   arch-check.sh    -> estrutura (AP-16: rota nao toca persistencia)
//   manual-tests.sh  -> comportamento via HTTP (precisa da app de pe)
//   internal-tests.js-> propriedades que a resposta HTTP nao revela
//
// O que so da para provar aqui:
//   1. Rollback da transacao do checkout. Nao da para fazer o SQLite falhar sob
//      demanda por curl, entao a falha e injetada em processo.
//   2. Que a checagem 1 sabe reprovar (controle negativo). Um teste que so sabe
//      passar nao mede nada.
//   3. Que src/app.js pode ser importado sem abrir porta — e propriedade do
//      modulo, nao de endpoint.
//   4. Que o teto de per_page e aplicado. Com 2 cursos no seed a resposta HTTP
//      fica identica com e sem limite, entao a prova tem que ser no validador.
//   5. Que o relatorio financeiro (AP-08) nao volta a ser N+1. O numero de
//      queries emitidas nao pode crescer com a quantidade de cursos.
//
// Uso:  node internal-tests.js     (ou: npm run test:internal)
// Saida: 0 = todas passaram, 1 = alguma falhou.

const { db, run, all, initSchema, withTransaction } = require('./src/database/connection');
const enrollmentModel = require('./src/models/enrollmentModel');
const reportService = require('./src/services/reportService');
const createApp = require('./src/app');
const { validateReportQuery } = require('./src/middlewares/validators');
const config = require('./src/config');

const results = [];

function check(name, passed, detail) {
    results.push({ name, passed, detail });
    console.log(`  ${passed ? '✓' : '✗'} ${name}`);
    if (detail) console.log(`      ${detail}`);
}

async function countEnrollments() {
    return (await all('SELECT id FROM enrollments')).length;
}

/** Conta quantas queries de leitura (db.all/db.get) o driver emite durante `work`. */
async function withQueryCount(work) {
    let count = 0;
    const originalAll = db.all.bind(db);
    const originalGet = db.get.bind(db);
    db.all = (...args) => { count += 1; return originalAll(...args); };
    db.get = (...args) => { count += 1; return originalGet(...args); };
    try {
        const result = await work();
        return { result, count };
    } finally {
        db.all = originalAll;
        db.get = originalGet;
    }
}

async function seedCourses(howMany, startingAt) {
    const ids = [];
    for (let i = 0; i < howMany; i += 1) {
        const title = `Curso ${startingAt + i}`;
        const { lastID: courseId } = await run(
            'INSERT INTO courses (title, price, active) VALUES (?, ?, 1)',
            [title, 100 + i]
        );
        const { lastID: enrollmentId } = await run(
            'INSERT INTO enrollments (user_id, course_id) VALUES (1, ?)',
            [courseId]
        );
        await run(
            "INSERT INTO payments (enrollment_id, amount, status) VALUES (?, ?, 'PAID')",
            [enrollmentId, 100 + i]
        );
        ids.push(courseId);
    }
    return ids;
}

/** 1. A transacao desfaz a matricula quando a gravacao seguinte falha. */
async function checkRollback() {
    const before = await countEnrollments();
    let propagated = null;

    try {
        await withTransaction(async () => {
            await enrollmentModel.create(1, 2);
            throw new Error('falha simulada gravando o pagamento');
        });
    } catch (err) {
        propagated = err.message;
    }

    const after = await countEnrollments();
    check(
        'rollback: matricula nao sobra quando o pagamento falha',
        after === before && propagated !== null,
        `erro propagado: ${propagated} | enrollments ${before} -> ${after}`
    );
}

/** 2. Controle negativo: sem transacao, a linha orfa TEM que sobrar. */
async function checkRollbackControl() {
    const before = await countEnrollments();

    try {
        await enrollmentModel.create(1, 2);
        throw new Error('falha simulada gravando o pagamento');
    } catch (err) {
        // esperado — o ponto e o que ficou no banco, nao o erro
    }

    const after = await countEnrollments();
    check(
        'controle negativo: sem transacao a linha orfa persiste (prova que a checagem acima mede algo)',
        after === before + 1,
        `enrollments ${before} -> ${after}`
    );
}

/** 3. A app monta sem efeito colateral — nada de porta aberta no import. */
function checkAppImportable() {
    const app = createApp();
    check(
        'app: src/app.js importavel sem abrir porta nem criar schema',
        typeof createApp === 'function' && typeof app === 'function',
        'quem chama listen() e o src/server.js'
    );
}

/** 4. per_page tem teto e default vindos do config. */
function checkPaginationClamp() {
    const excessivo = { query: { per_page: '999999999' } };
    validateReportQuery(excessivo, null, () => {});

    const semParametro = { query: {} };
    validateReportQuery(semParametro, null, () => {});

    check(
        'paginacao: per_page limitado ao teto e default vindo do config',
        excessivo.pagination.perPage === config.maxPageSize &&
            semParametro.pagination.perPage === config.defaultPageSize,
        `per_page=999999999 -> ${excessivo.pagination.perPage} (teto ${config.maxPageSize}) | ` +
            `sem per_page -> ${semParametro.pagination.perPage}`
    );
}

/** 5. financialReport (AP-08): numero de queries constante, nao cresce com N cursos. */
async function checkFinancialReportQueryCountConstant() {
    await seedCourses(3, 100);
    const { count: smallCount, result: smallReport } = await withQueryCount(() =>
        reportService.financialReport({ page: 1, perPage: 50 })
    );

    await seedCourses(12, 200);
    const { count: largeCount, result: largeReport } = await withQueryCount(() =>
        reportService.financialReport({ page: 1, perPage: 50 })
    );

    check(
        'financialReport: numero de queries nao cresce com a quantidade de cursos',
        smallCount === largeCount && largeCount <= 3,
        `${smallReport.length} curso(s) -> ${smallCount} querie(s) | ` +
            `${largeReport.length} curso(s) -> ${largeCount} querie(s)`
    );
}

/** 6. Controle negativo: uma query por curso (jeito antigo) TEM que crescer com N. */
async function checkFinancialReportControlWouldGrow() {
    const courses = await all('SELECT id FROM courses');

    const { count: naiveCount } = await withQueryCount(async () => {
        for (const course of courses) {
            await enrollmentModel.findDetailsByCourseIds([course.id]);
        }
    });

    check(
        'controle negativo: uma query por curso cresce com N (prova que a checagem acima mede algo)',
        naiveCount === courses.length,
        `${courses.length} curso(s) -> ${naiveCount} querie(s) no jeito ingenuo`
    );
}

(async () => {
    console.log('==========================================================');
    console.log('internal-tests — ecommerce-api-legacy');
    console.log('Checagens em processo (sem servidor)');
    console.log('==========================================================\n');

    await initSchema();

    await checkRollback();
    await checkRollbackControl();
    checkAppImportable();
    checkPaginationClamp();
    await checkFinancialReportQueryCountConstant();
    await checkFinancialReportControlWouldGrow();

    const failed = results.filter((r) => !r.passed);
    console.log('\n==========================================================');
    if (failed.length === 0) {
        console.log(`PASS: ${results.length} checagem(ns) internas OK.`);
        console.log('==========================================================');
        process.exit(0);
    }
    console.log(`FAIL: ${failed.length} de ${results.length} checagem(ns) falharam.`);
    failed.forEach((r) => console.log(`  - ${r.name}`));
    console.log('==========================================================');
    process.exit(1);
})();
